File: basic_dsa_ninety_two.java
/**
 * Basic DSA Problems - Reverse Words in a String III
 * Simple and essential DSA problems for beginners
 */
public class basic_dsa_ninety_two {

    // Problem: Reverse the order of characters in each word within a sentence (words separated by spaces).
    // Example: "Let's take LeetCode contest" -> "s'teL ekat edoCteeL tsetnoc"

    /**
     * Split by space, reverse each word, join.
     * Time O(n), Space O(n)
     */
    public static String reverseWords(String s) {
        if (s == null) return null;
        String[] words = s.split(" ");
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < words.length; i++) {
            if (i > 0) sb.append(" ");
            sb.append(reverse(words[i]));
        }
        return sb.toString();
    }

    private static String reverse(String w) {
        char[] a = w.toCharArray();
        int i = 0, j = a.length - 1;
        while (i < j) {
            char t = a[i];
            a[i] = a[j];
            a[j] = t;
            i++;
            j--;
        }
        return new String(a);
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Reverse Words in a String III");
        System.out.println("=".repeat(60));

        System.out.println("Let's take LeetCode contest -> " + reverseWords("Let's take LeetCode contest"));
    }
}

File: basic_dsa_ninety_two.cpp
/**
 * Basic DSA - Shuffle the Array (LeetCode 1470)
 * Shuffle [x1,x2,...,xn,y1,y2,...,yn] to [x1,y1,x2,y2,...,xn,yn].
 */
#include <iostream>
#include <vector>

std::vector<int> shuffle(std::vector<int>& nums, int n) {
    std::vector<int> out(2 * n);
    for (int i = 0; i < n; i++) {
        out[2 * i] = nums[i];
        out[2 * i + 1] = nums[n + i];
    }
    return out;
}

void print(std::vector<int>& v) {
    std::cout << "[";
    for (size_t i = 0; i < v.size(); i++) {
        std::cout << v[i];
        if (i + 1 < v.size()) std::cout << ",";
    }
    std::cout << "]\n";
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Shuffle the Array\n";
    std::cout << std::string(60, '=') << "\n";

    std::vector<int> a1 = {2, 5, 1, 3, 4, 7};
    auto r1 = shuffle(a1, 3);
    std::cout << "[2,5,1,3,4,7], n=3 -> ";
    print(r1);

    std::vector<int> a2 = {1, 2, 3, 4, 4, 3, 2, 1};
    auto r2 = shuffle(a2, 4);
    std::cout << "[1,2,3,4,4,3,2,1], n=4 -> ";
    print(r2);
}
